posydon/utils/posydonwarning.py

```diff
@@ -77,33 +105,336 @@ def __init__(self, message=''):
+    """Filter a warning.
+
+        Parameters
+        ----------
+        warning : dict (default: {message="No warning"})
+            Dictionary containing all options passed to warnings.warn.
+
+        Returns
+        -------
+        warning or None in case it got filtered out.
+    """
+    global _POSYDON_WARNINGS_REGISTRY
+    if not isinstance(warning, dict):
+        raise TypeError("warning must be a dictionary.")
+    # get stack level
+    stacklevel = warning.get('stacklevel', 0)
+    if not isinstance(stacklevel, int):
+        raise TypeError("stacklevel must be an integer.")
+    # get category
+    category = warning.get('category', None)
+    if not(isinstance(category, type) and issubclass(category, Warning)):
+        category = _get_POSYDONWarning_class(category)
+    if not isinstance(category, type):
+        category = UserWarning
+    # get filename and lineno from frame at stacklevel
+    try:
+        frame = sys._getframe(stacklevel)
+    except:
+        g = sys.__dict__
+        filename = "sys"
+        lineno = 1
+    else:
+        g = frame.f_globals
+        filename = frame.f_code.co_filename
+        lineno = frame.f_lineno
+    # get module
+    if isinstance(g, dict):
+        module = g.get('__name__', "posydonwarnings")
+    else:
+        module = "posydonwarnings"
+    # check registry
+    if not isinstance(_POSYDON_WARNINGS_REGISTRY, dict):
+        print("Reset _POSYDON_WARNINGS_REGISTRY, old was:",
+              _POSYDON_WARNINGS_REGISTRY)
+        _POSYDON_WARNINGS_REGISTRY = {}
+    # set key for registry
+    key = (category, filename, lineno)
+    # get message text
+    text = warning.get('message', "")
+    if not isinstance(text, str):
+        raise TypeError("message must be an integer.")
+    # search the filters
+    for item in warnings.filters:
+        action, msg, cat, mod, ln = item
+        if ((msg is None or msg.match(text)) and
+            issubclass(category, cat) and
+            (mod is None or mod.match(module)) and
+            (ln == 0 or lineno == ln)):
+            break
+    else:
+        action = "default"
+    # apply action
+    if action == "ignore":
+        return None
+    elif action in ["always", "default", "module", "once"]:
+        if key in _POSYDON_WARNINGS_REGISTRY:
+            _POSYDON_WARNINGS_REGISTRY[key] += 1
+            if action == "default":
+                return None
+        else:
+            _POSYDON_WARNINGS_REGISTRY[key] = 1
+    return warning
+
+def _issue_warn(warning=dict(message="No warning")):
+    """Issue a warning.
+
+        Parameters
+        ----------
+        warning : dict (default: {message="No warning"})
+            Dictionary containing all options passed to warnings.warn.
+    """
+    filtered_warning = _apply_POSYDON_filter(warning)
+    if filtered_warning is None:
+        return
+    else:
+        warnings.warn(**filtered_warning)
+
+
+class _Catched_POSYDON_Warnings:
```

Comment:
A general comment, it is unclear to me how much of these new functions are *essentially* the same as the python source code for catch_warnings, and what changes introduce new functionality. I think it would be _very_ helpful, if there are only a few minor changes in functionality, to highlight exactly where those are in the code itself so that people understand the differences here. I know that you wrote a couple sentences in the docs at the top, but at least for me, that is not enough to really understand what was changed from the python source code, without having to go look at the python source code and compare.

Reply:
I added more comments, please let me know, if there is need for more (tell please where you'd need more).

Reply:
Ok I think the additional comments are helpful!